specs/DragStarting.md

```diff
@@ -0,0 +1,196 @@
+interface ICoreWebView2StagingDragStartingEventArgs : IUnknown {
+  /// The operations this drag data supports.
+  [propget] HRESULT DragAllowedOperations([out, retval] COREWEBVIEW2_DRAG_EFFECTS* value);
+
+
+  /// The data being dragged.
+  [propget] HRESULT DragData([out, retval] IDataObject** value);
+
+  /// The position at which drag was detected.
+  [propget] HRESULT DragPosition([out, retval] POINT* value);
+
+
+  /// Gets the `Handled` property.
+  [propget] HRESULT Handled([out, retval] BOOL* value);
+
+
+  /// Indicates whether this event has been handled by the app.  If the
+  /// app handles this event, WebView2 will not initiate drag drop.  If
+  /// the app does not handle the event, WebView2 will initiate its own
+  /// drag drop logic.
+  [propput] HRESULT Handled([in] BOOL value);
+
+
+
+  /// Returns an `ICoreWebView2Deferral` object. Use this operation to complete
+  /// the CoreWebView2DragStartingEventArgs.
+  ///
+  /// Until the deferral is completed, subsequent attempts to initiate drag
+  /// in the WebView2 will fail and if the cursor was changed as part of
+  /// drag it will not restore.
+  HRESULT GetDeferral(
+      [out, retval] ICoreWebView2Deferral** value);
+
+
+}
+
+/// Receives `DragStarting` events.
+[uuid(3b149321-83c3-5d1f-b03f-a42899bc1c15), object, pointer_default(unique)]
+interface ICoreWebView2StagingDragStartingEventHandler : IUnknown {
+  /// Provides the event args for the corresponding event.
+  HRESULT Invoke(
+      [in] ICoreWebView2CompositionController* sender,
+      [in] ICoreWebView2StagingDragStartingEventArgs* args);
+}
+
+/// A continuation of the ICoreWebView2CompositionController4 interface.
+/// This interface includes an API which exposes the DragStarting event.
+[uuid(975d6824-6a02-5e98-ab7c-e4679d5357f4), object, pointer_default(unique)]
+interface ICoreWebView2StagingCompositionController : IUnknown {
+  /// Adds an event handler for the `DragStarting` event.
+  /// Adds an event handler for the `DragStarting` event.  `DragStarting` is
+  /// raised when the WebView2 detects a drag started within the WebView2.
+  /// This event can be used to override WebView2's default drag starting
+  /// logic.
+  HRESULT add_DragStarting(
+      [in] ICoreWebView2StagingDragStartingEventHandler* eventHandler,
+      [out] EventRegistrationToken* token);
+
+  /// Removes an event handler previously added with `add_DragStarting`.
+  HRESULT remove_DragStarting(
+      [in] EventRegistrationToken token);
+
+
+}
+```
+```c# (but really MIDL3)
+    runtimeclass CoreWebView2DragStartingEventArgs
+    {
+
+        Windows.ApplicationModel.DataTransfer.DataPackageOperation DragAllowedOperations { get; };
+
+        Windows.ApplicationModel.DataTransfer.DataPackage DragData { get; };
+
+        Windows.Foundation.Point DragPosition { get; };
+
+        Boolean Handled { get; set; };
+
+
+        Windows.Foundation.Deferral GetDeferral();
+
```

Comment:
The WinUI XAML version of this event has a few more properties than ours. Is that fine?

https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.drageventargs.acceptedoperation?view=windows-app-sdk-1.5

Reply:
Yes it is fine. Xaml exposes more options/state than is strictly necessary to perform a drag.